File: BaekJoon/Stack/_Silver/압축_1662.cpp
//
//  압축_1662.cpp
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/03/03.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: - 압축 1662
// MARK: Stack C++ 문제풀이
//1 + 3 * (2 + 2 * (1 + 1 * 1))
// ★ depth에 따른 값 누적 및 연산 수행으로 값 도출

#if 0
#include <string>
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

int main() {
    ios_base :: sync_with_stdio(0); cin.tie(0);
    string S; cin>>S;
    stack<int> STK;
    vector<int> depth(51,0);
    for(int i=0; i<S.length(); i++) {
        if(S[i]=='(') {
            STK.push(i);
        } else if(S[i]==')') {
            int last = (int)STK.top();
            depth[(int)STK.size()-1] += (S[last-1]-'0') * depth[(int)STK.size()] - 1;
            depth[(int)STK.size()] = 0;
            STK.pop();
        } else {
            depth[(int)STK.size()]++;
        }
    }
    
    printf("%d\n",depth[0]);
    return 0;
}
#endif

File: BaekJoonPS/BaekjoonAlgorithmPS/Stack/_Silver/압축_1662.swift
//
//  압축_1662.swift
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/03/04.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: - 압축 1662

// MARK: Stack swift 복습 문제풀이 (최적화)
/*
let S = Array(readLine()!)
var idx = 0
var G = [[Character]](repeating: [Character](), count: 101)
var V = [Int](repeating: 0, count: 101)

for i in S.indices {
  if S[i] == "(" { idx += 1 }
  else if S[i] == ")" {
    idx -= 1
    V[idx] += Int(G[idx].last!.asciiValue! - 48) * V[idx+1] - 1
    V[idx+1] = 0
    G[idx+1] = []
  } else {
    V[idx] += 1
    G[idx].append(S[i])
  }
}

print(V[0])
 */


// MARK: Stack swift 문제풀이

/*
import Foundation

let arr = Array(readLine()!)
var STK = [Int]()
var G = [Int](repeating: 0, count: 51)

for i in arr.indices {
    if arr[i] == "(" {
        STK.append(i)
    } else if arr[i] == ")" {
        let depth = STK.count
        let topIdx = STK.last!
        G[depth-1] += (Int(arr[topIdx-1].unicodeScalars.first!.value - 48) * G[depth]) - 1
        STK.removeLast()
        G[depth] = 0
    } else {
        let depth = STK.count
        G[depth] += 1
    }
}

print(G[0])
*/
